pkg/charm/path.go

```diff
@@ -0,0 +1,37 @@
+package charm
+
+import (
+	"fmt"
+	"strings"
+)
+
+type Path []*instance
+
+func (c Path) run(args []string) error {
+	err := c.last().command.Run(args)
+	if err == ErrNoRun {
+		err = fmt.Errorf("%q: a sub-command is required: %s", c.pathname(args...), c.subCommands())
+	}
+	return err
+}
+
+func (c Path) last() *instance {
+	return c[len(c)-1]
+}
+
+func (c Path) pathname(args ...string) string {
+	names := make([]string, 0, len(c))
```

Comment:
(But I still like `var name []string` in this situation for lower cognitive load.)
```suggestion
	names := make([]string, 0, len(c)+len(args))
```

Reply:
I prefer using one pattern for when you know the size of slices (for lower cognitive load) rather than switching between the patterns based on performance considerations.  This is the one Go-ism that I think you and I will continue to disagree on :-)

Reply:
And that's OK!